p2p/gossip/protocol.go

```diff
@@ -2,405 +2,340 @@ package gossip
-			return
-		}
+	if msg.Metadata.ClientVersion != protocolVer {
+		prot.Log.Error("fail to validate message's protocol version when validating gossip message, err %v", err)
+		return err
 	}
-
-}
-
-// Close closes the neighborhood and shutsdown requests for more peers
-func (s *Neighborhood) Close() {
-	// no need to shutdown con, conpool will do so in a shutdown. the morepeerreq won't work
-	close(s.shutdown)
-	//todo close all peers
+	return nil
 }
 
 // Broadcast is the actual broadcast procedure, loop on peers and add the message to their queues
-func (s *Neighborhood) Broadcast(msg []byte) error {
-
-	if len(s.peers)+len(s.inpeers) == 0 {
-		return errors.New("No peers in neighborhood")
+func (prot *Protocol) Broadcast(payload []byte, nextProt string) error {
+	// add gossip header
+	header := &pb.Metadata{
+		NextProtocol:  nextProt,
+		ClientVersion: protocolVer,
+		Timestamp:     time.Now().Unix(),
+		AuthPubKey:    prot.signer.PublicKey().Bytes(),
+		MsgSign:       nil,
 	}
 
-	oldmessage := false
+	msg := &pb.ProtocolMessage{
+		Metadata: header,
+		Data:     &pb.ProtocolMessage_Payload{Payload: payload},
+	}
 
-	checksum := generateID(msg)
+	bin, err := proto.Marshal(msg)
+	if err != nil {
+		prot.Log.Error("failed to marshal message when generating gossip header, err %v", err)
+		return err
 
-	s.oldMessageMu.RLock()
-	if _, ok := s.oldMessageQ[checksum]; ok {
-		// todo : - have some more metrics for termination
-		// todo	: - maybe tell the peer weg ot this message already?
-		oldmessage = true
 	}
-	s.oldMessageMu.RUnlock()
 
-	if !oldmessage {
-		s.oldMessageMu.Lock()
-		s.oldMessageQ[checksum] = struct{}{}
-		s.oldMessageMu.Unlock()
+	sign, err2 := prot.signer.Sign(bin)
+	if err2 != nil {
+		prot.Log.Error("failed to Sign header when generating gossip header, err %v", err)
+		return err
 	}
 
-	s.peersMutex.RLock()
-	for p := range s.peers {
-		peer := s.peers[p]
-		go peer.addMessage(msg, checksum)
-		s.Debug("adding message to peer %v", peer.Pretty())
-	}
-	s.peersMutex.RUnlock()
-	s.inpeersMutex.RLock()
-	for p := range s.inpeers {
-		peer := s.inpeers[p]
-		go peer.addMessage(msg, checksum)
-		s.Debug("adding message to peer %v", peer.Pretty())
-	}
-	s.inpeersMutex.RUnlock()
+	msg.Metadata.MsgSign = sign
 
-	//TODO: if we didn't send to RandomConnections then try to other peers.
-	if oldmessage {
-		return errors.New("old message")
+	finbin, err := proto.Marshal(msg)
+	if err != nil {
+		return err
 	}
 
+	// so we won't process our own messages
+	hash := calcHash(finbin)
+	prot.markMessage(hash)
+	prot.propagateMessage(finbin, hash)
 	return nil
 }
 
-// getMorePeers tries to fill the `peers` slice with dialed outbound peers that we selected from the dht.
-func (s *Neighborhood) getMorePeers(numpeers int) int {
+// Start a loop that process peers events
+func (prot *Protocol) Start() {
+	prot.peerConn, prot.peerDisc = prot.net.SubscribePeerEvents()
```

Comment:
either put this in `NewProtocol` or insert this to the eventloop and remove the fields from `Protocol`